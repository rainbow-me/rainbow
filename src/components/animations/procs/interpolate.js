import Animated from 'react-native-reanimated';

const interpolateTwo = Animated.proc(
  (value, extrapolate, input1, input2, output1, output2) =>
    Animated.interpolateNode(value, {
      extrapolate,
      inputRange: [input1, input2],
      outputRange: [output1, output2],
    })
);

const interpolateThree = Animated.proc(
  (value, extrapolate, input1, input2, input3, output1, output2, output3) =>
    Animated.interpolateNode(value, {
      extrapolate,
      inputRange: [input1, input2, input3],
      outputRange: [output1, output2, output3],
    })
);

const interpolateFour = Animated.proc(
  (
    value,
    extrapolate,
    input1,
    input2,
    input3,
    input4,
    output1,
    output2,
    output3,
    output4
  ) =>
    Animated.interpolateNode(value, {
      extrapolate,
      inputRange: [input1, input2, input3, input4],
      outputRange: [output1, output2, output3, output4],
    })
);

const interpolateFive = Animated.proc(
  (
    value,
    extrapolate,
    input1,
    input2,
    input3,
    input4,
    input5,
    output1,
    output2,
    output3,
    output4,
    output5
  ) =>
    Animated.interpolateNode(value, {
      extrapolate,
      inputRange: [input1, input2, input3, input4, input5],
      outputRange: [output1, output2, output3, output4, output5],
    })
);

export default function interpolate(
  value,
  { extrapolate = Animated.Extrapolate.EXTEND, inputRange, outputRange } = {}
) {
  if (inputRange.length === 2) {
    return interpolateTwo(value, extrapolate, ...inputRange, ...outputRange);
  }
  if (inputRange.length === 3) {
    return interpolateThree(value, extrapolate, ...inputRange, ...outputRange);
  }
  if (inputRange.length === 4) {
    return interpolateFour(value, extrapolate, ...inputRange, ...outputRange);
  }
  if (inputRange.length === 5) {
    return interpolateFive(value, extrapolate, ...inputRange, ...outputRange);
  }
}
