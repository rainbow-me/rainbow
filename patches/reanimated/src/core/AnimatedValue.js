import { createAnimatedSet as set } from './AnimatedSet';
import interpolate from '../derived/interpolate';
import InternalAnimatedValue from './InternalAnimatedValue';
import { Platform } from 'react-native';
import { evaluateOnce } from '../derived/evaluateOnce';
import ReanimatedModule from '../ReanimatedModule';
import { val } from '../val';

// Animated value wrapped with extra methods for omit cycle of dependencies
export default class AnimatedValue extends InternalAnimatedValue {
  setValue(value) {
    this.__detachAnimation(this._animation);
    if (Platform.OS === 'web' || Platform.OS === 'windows' || Platform.OS === 'macos') {
      this._updateValue(val(value));
    } else {
      if (ReanimatedModule.setValue && typeof value === 'number') {
        // FIXME Remove it after some time
        // For OTA-safety
        // FIXME handle setting value with a node
        ReanimatedModule.setValue(this.__nodeID, value);
      } else {
        evaluateOnce(set(this, value), this);
      }
    }
  }

  toString() {
    return `AnimatedValue, id: ${this.__nodeID}`;
  }

  interpolate(config) {
    return interpolate(this, config);
  }
}
