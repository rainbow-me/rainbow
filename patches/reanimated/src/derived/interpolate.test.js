import interpolate from './interpolate';
import AnimatedValue from '../core/AnimatedValue';

jest.mock('../ReanimatedEventEmitter');
jest.mock('../ReanimatedModule');

const value = new AnimatedValue(0);

it('throws if inputRange or outputRange does not contain at least 2 elements', () => {
  expect(() =>
    interpolate(value, {
      inputRange: [0],
      outputRange: [0, 1],
    })
  ).toThrowErrorMatchingSnapshot();
  expect(() =>
    interpolate(value, {
      inputRange: [0, 1],
      outputRange: [0],
    })
  ).toThrowErrorMatchingSnapshot();
});

it('throws if inputRange and outputRange are not the same length', () => {
  expect(() =>
    interpolate(value, {
      inputRange: [0, 1, 2],
      outputRange: [0, 1, 2, 3],
    })
  ).toThrowErrorMatchingSnapshot();
});

it('throws if inputRange or outputRange contains an invalid value', () => {
  expect(() =>
    interpolate(value, {
      inputRange: [0, 1, Infinity],
      outputRange: [0, 1, 2],
    })
  ).toThrowErrorMatchingSnapshot();
  expect(() =>
    interpolate(value, {
      inputRange: [0, 1, 2],
      outputRange: [0, 1, NaN],
    })
  ).toThrowErrorMatchingSnapshot();
});

it('throws if inputRange is not monotonically non-decreasing', () => {
  expect(() =>
    interpolate(value, {
      inputRange: [0, 1, 0],
      outputRange: [0, 1, 2],
    })
  ).toThrowErrorMatchingSnapshot();
});
