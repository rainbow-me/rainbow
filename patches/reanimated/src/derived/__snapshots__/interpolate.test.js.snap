// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws if inputRange and outputRange are not the same length 1`] = `"inputRange and outputRange must be the same length."`;

exports[`throws if inputRange is not monotonically non-decreasing 1`] = `"inputRange must be monotonically non-decreasing. (0,1,0)"`;

exports[`throws if inputRange or outputRange contains an invalid value 1`] = `"inputRange cannot include Infinity. (0,1,Infinity)"`;

exports[`throws if inputRange or outputRange contains an invalid value 2`] = `"outputRange cannot include NaN. (0,1,NaN)"`;

exports[`throws if inputRange or outputRange does not contain at least 2 elements 1`] = `"inputRange must have at least 2 elements. (0)"`;

exports[`throws if inputRange or outputRange does not contain at least 2 elements 2`] = `"outputRange must have at least 2 elements. (0)"`;
